docs/docs/configuration-parameters.md
---
title: Mesos-DNS Configuration Parameters
---

# Mesos-DNS Configuration Parameters

Mesos-DNS is configured through the parameters in a JSON file. You can point Mesos-DNS to a specific configuration file using the argument `-config=pathto/file.json`. If no configuration file is passed as an argument, Mesos-DNS will look for file `config.json` in the current directory.

The configuration file should include the following fields:

```json
{
  "zk": "zk://10.101.160.15:2181/mesos",
  "masters": ["10.101.160.15:5050", "10.101.160.16:5050", "10.101.160.17:5050"],
  "mesosCredentials": {
    "principal": "my-mesos-dns-principal",
    "secret": "super-secret"
  },
  "mesosAuthentication": "basic",
  "refreshSeconds": 60,
  "ttl": 60,
  "domain": "mesos",
  "port": 53,
  "resolvers": ["169.254.169.254"],
  "zoneResolvers": {
    "weave": ["172.17.0.1"]
  },
  "timeout": 5,
  "httpon": true,
  "dnson": true,
  "httpport": 8123,
  "externalon": true,
  "listener": "10.101.160.16",
  "SOAMname": "ns1.mesos",
  "SOARname": "root.ns1.mesos",
  "SOARefresh": 60,
  "SOARetry":   600,
  "SOAExpire":  86400,
  "SOAMinttl": 60,
  "IPSources": ["netinfo", "mesos", "host"]
}
```

`zk` is a link to the Zookeeper instances on the Mesos cluster. Its format is `zk://host1:port1,host2:port2/mesos/`, where the number of hosts can be one or more. The default port for Zookeeper is `2181`. Mesos-DNS will monitor the Zookeeper instances to detect the current leading master.

`zkDetectionTimeout` defines how long to wait (in seconds) for Zookeeper to report a new leading Mesos master. This timeout is activated on:

- Start up, where it plays the role of the "initial leader detection timeout" via ZK.
- Mesos cluster changes, where there is no leading master for some period of time.
- Zookeeper or network failure, when losing connection to the ZK cluster.

If a *non-zero* timeout is specified and the timeout threshold is exceeded before
a new leading Mesos master is reported by the ZK-based master detector, the program will exit.

Defaults to `30` seconds.

`masters` is a comma separated list with the IP address and port number for the master(s) in the Mesos cluster. Mesos-DNS will automatically find the leading master at any point in order to retrieve state about running tasks. If there is no leading master or the leading master is not responsive, Mesos-DNS will continue serving DNS requests based on stale information about running tasks. The `masters` field is required.

It is sufficient to specify just one of the `zk` or `masters` field. If both are defined, Mesos-DNS will first attempt to detect the leading master through Zookeeper. If Zookeeper is not responding, it will fall back to using the `masters` field. Both `zk` and `master` fields are static. To update them you need to restart Mesos-DNS. We recommend you use the `zk` field since this allows the dynamic addition to Mesos masters.

`mesosAuthentication` configures the authentication mechanism for talking to the Mesos cluster. Valid values are '', 'basic' (see `mesosCredentials`), and 'iam'. Default is ''.

`mesosCredentials` is a dictionary containing a `principal` and a `secret`, corresponding to a configured authentication principal for the Mesos masters. Starting with Mesos `1.0.0`, if the masters have `http_authentication` enabled, then Mesos-DNS must authenticate. You must specify `mesosAuthentication`: `basic` to use this configuration.

`refreshSeconds` is the frequency at which Mesos-DNS updates DNS records based on information retrieved from the Mesos master. The default value is 60 seconds.

`stateTimeoutSeconds` is the time that Mesos-DNS will wait for the Mesos master to respond to its request for `state.json` in seconds. The default value is 300 seconds.

`ttl` is the [time to live](http://en.wikipedia.org/wiki/Time_to_live#DNS_records) value for DNS records served by Mesos-DNS, in seconds. It allows caching of the DNS record for a period of time in order to reduce DNS request rate. `ttl` should be equal or larger than `refreshSeconds`. The default value is 60 seconds.

`domain` is the domain name for the Mesos cluster. The domain name can use characters [a-z, A-Z, 0-9], `-` if it is not the first or last character of a domain portion, and `.` as a separator of the textual portions of the domain name. We recommend you avoid valid [top-level domain names](http://en.wikipedia.org/wiki/List_of_Internet_top-level_domains). The default value is `mesos`.

`port` is the port number that Mesos-DNS monitors for incoming DNS requests. Requests can be sent over TCP or UDP. We recommend you use port `53` as several applications assume that the DNS server listens to this port. The default value is `53`.

`resolvers` is a comma separated list with the IP addresses of external DNS servers that Mesos-DNS will contact to resolve any DNS requests outside the `domain`. We ***recommend*** that you list the nameservers specified in the `/etc/resolv.conf` on the server Mesos-DNS is running. Alternatively, you can list `8.8.8.8`, which is the [Google public DNS](https://developers.google.com/speed/public-dns/) address. The `resolvers` field is required.

`zoneResolvers` is a dictionary of zone-specific external DNS servers, where the key is the matching zone (sans leading / trailing .). You can use this configuration option to route a subset of DNS queries to a specific set of DNS servers. Note, general, catch-all resolvers are still specified with `resolvers`.

`timeout` is the timeout threshold, in seconds, for connections and requests to external DNS requests. The default value is 5 seconds.

`listener` is the IP address of Mesos-DNS. In SOA replies, Mesos-DNS identifies hostname `mesos-dns.domain` as the primary nameserver for the domain. It uses this IP address in an A or AAAA record for `mesos-dns.domain`. The default value is "0.0.0.0", which instructs Mesos-DNS to create an A record for every IP address associated with a network interface on the server that runs the Mesos-DNS process.

`dnson` is a boolean field that controls whether Mesos-DNS listens for DNS requests or not. The default value is `true`.

`httpon` is a boolean field that controls whether Mesos-DNS listens for HTTP requests or not. The default value is `true`.

`httpport` is the port number that Mesos-DNS monitors for incoming HTTP requests. The default value is `8123`.

`externalon` is a boolean field that controls whether Mesos-DNS serves requests outside of the Mesos domain. The default value is `true`.

`srvRecordDefaultWeight` is an int field field that defines the default weight given to SRV records. The default value is `1`.

`SOAMname` specifies the domain name of the name server that was the original or primary source of data for the configured domain.
The configured name will always be converted to a FQDN by ensuring it ends with a `.`. The default value is `ns1.mesos`.

`SOARname` specifies the mailbox of the person responsible for the configured domain. The format is `mailbox.domain`, using a `.` instead of `@`. i.e. `root@ns1.mesos` becomes `root.ns1.mesos`. For details, see the [RFC-1035](http://tools.ietf.org/html/rfc1035#page-18). The default value is `root.ns1.mesos`.

`SOARefresh` is the REFRESH field in the SOA record for the Mesos domain. For details, see the [RFC-1035](http://tools.ietf.org/html/rfc1035#page-18). The default value is `60`.

`SOARetry` is the RETRY field in the SOA record for the Mesos domain. For details, see the [RFC-1035](http://tools.ietf.org/html/rfc1035#page-18). The default value is `600`.

`SOAExpire` is the EXPIRE field in the SOA record for the Mesos domain. For details, see the [RFC-1035](http://tools.ietf.org/html/rfc1035#page-18). The default value is `86400`.

`SOAMinttl` is the minimum TTL field in the SOA record for the Mesos domain. For details, see the [RFC-2308](https://tools.ietf.org/html/rfc2308). The default value is `60`.

`recurseon` controls if the DNS replies for names in the Mesos domain will indicate that recursion is available. The default value is `true`.

`enforceRFC952` will enforce an older, more strict set of rules for DNS labels. For details, see the [RFC-952](https://tools.ietf.org/html/rfc952). The default value is `false`.

`IPSources` defines a fallback list of IP sources for task records,
sorted by priority. If you use **Docker**, and enable the `netinfo` IPSource, it may cause tasks to become unreachable, because after Mesos 0.25, the Docker executor publishes the container's internal IP in NetworkInfo. The default value is: `["netinfo", "mesos", "host"]`

- `host`: Host IP of the Mesos slave where a task is running.
- `mesos`: Mesos containerizer IP. **DEPRECATED**
- `docker`: Docker containerizer IP. **DEPRECATED**
- `netinfo`: Mesos 0.25 NetworkInfo.


CHANGELOG
2016-09-14: v0.6.0
  Added some benchmarks for RecordGenerator.insertRR
  Make minor fixes to enumeration API, where it would return null, as
    as opposed to an empty list
  Add the AXFR (record listing) HTTP API endpoint
  Make HTTP listener address configurable
  General improvements to the HTTP client
    -Pool HTTP connections, preventing connection leakage
    -Set the User Agent
  Switch the project to Go Vendor as opposed to Godeps
  Add HTTPS Client support to Mesos-DNS (for state.json)
  Add Mesos authentication support
  Fix tasks from more than 63 dashes and a non - character from
    crashing Mesos-DNS
  Don't treat truncated records as an error.

2016-03-04: v0.5.2
  Add EnforceRFC952 to sample config
  Improve performance of record generation
  Fix bug in DiscoveryInfo to use the same scheme for record creation
    as normal records (#377)
  Add enumeration API (#402)
  Generate additional SRV records if DiscoveryInfo port name is specified

2015-11-25: v0.5.1
  Add code signing to CI portion of repo
  documentation: Try to alleviate issues around IPSources
  resolver: Fix errant NXDOMAIN errors 
  
2015-11-19: v0.5.0
  Move repo, CI, and build tools to use Go 1.5
  resolver: Remove answers to fit 512 byte when truncating (#237, #326)
  generator: Add Timeout when fetching state.json (#243)
  generator: Fix hashString to introduce a sha1 hash (#157)
  Add support for Mesos 0.26 "NetworkInfo"
    -If you use NetworkInfo, in order to upgrade from Mesos 0.25->Mesos 0.26
     you must use Mesos-DNS 0.5.0

2015-10-09: v0.4.0
  Smarter leading Mesos master detection timeout logic (#284, #306, #320)
  resolver: Fix data races to RNG and SOASerial (#317)
  Remove bogus "NS request" error log line (#316)
  Transparent forwarding as per RFC5625 (#301, #307)
  Idiomatic struct definitions (#232, #312)
  exchanger: Improve error logging (#303)
  Generate correct SOA and NS records (#295, #296)
  detect: Extract address from MasterInfo.Address (#289, #290)
  state: Call given src only once per statusIPs call (#285)

2015-09-23: v0.3.0
  Add support for Mesos 0.25 "NetworkInfo" as an IPSource (#266, #267, #268)
  Find home directory with environment variables (#275, #277, #278)

2015-09-15: v0.2.0
  Mesos 0.24.0 compatibility (#263, #228)
  Configurable ZooKeeper detection timeout (#260)
  Configurable "IPSources" for A and SRV records (#226)
  Mesos framework A and SRV records (#209)
  Mesos slave A and SRV records (#208)
  Mesos task container A records (#204)
  Default to RFC1123 labels (#177)
  Compress all DNS replies (#173)
  Mesos master A and SRV records (#169)
  Resolver and masters configuration validation (#158)
  Fix buggy character range used for RFC952 label conversion (#155)
  Patch up slave address resolution (#153, #151)


records/config.go
package records

import (
	"crypto/tls"
	"crypto/x509"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/mesosphere/mesos-dns/errorutil"
	"github.com/mesosphere/mesos-dns/httpcli"
	"github.com/mesosphere/mesos-dns/httpcli/basic"
	"github.com/mesosphere/mesos-dns/httpcli/iam"
	"github.com/mesosphere/mesos-dns/logging"
	"github.com/miekg/dns"
)

// Config holds mesos dns configuration
type Config struct {
	// Refresh frequency: the frequency in seconds of regenerating records (default 60)
	RefreshSeconds int
	// Resolver port: port used to listen for slave requests (default 53)
	Port int
	// Timeout is the default connect/read/write timeout for outbound
	// queries
	Timeout int
	// Timeout in seconds waiting for the master to return data from StateJson
	StateTimeoutSeconds int
	// Zookeeper Detection Timeout: how long in seconds to wait for Zookeeper to
	// be initially responsive. Default is 30 and 0 means no timeout.
	ZkDetectionTimeout int
	// NOTE(tsenart): HTTPPort, DNSOn and HTTPOn have defined JSON keys for
	// backwards compatibility with external API clients.
	HTTPPort int `json:"HttpPort"`
	// TTL: the TTL value used for SRV and A records (default 60)
	TTL int32
	// SRVRecordDefaultWeight: default weight for SRV records
	SRVRecordDefaultWeight uint16
	// SOA record fields (see http://tools.ietf.org/html/rfc1035#page-18)
	SOASerial  uint32 // initial version number (incremented on refresh)
	SOARefresh uint32 // refresh interval
	SOARetry   uint32 // retry interval
	SOAExpire  uint32 // expiration time
	SOAMinttl  uint32 // minimum TTL
	SOAMname   string // primary name server
	SOARname   string // email of admin esponsible
	// Mesos master(s): a list of IP:port pairs for one or more Mesos masters
	Masters []string
	// DNS server: IP address of the DNS server for forwarded accesses
	ZoneResolvers map[string][]string
	// DNS server: a list of IP addresses or IP:port pairs for DNS servers for forwarded accesses
	Resolvers []string
	// IPSources is the prioritized list of task IP sources
	IPSources []string // e.g. ["host", "docker", "mesos", "rkt"]
	// Zookeeper: a single Zk url
	Zk string
	// Domain: name of the domain used (default "mesos", ie .mesos domain)
	Domain string
	// File is the location of the config.json file
	File string
	// Listen is the server DNS listener IP address
	Listener string
	// HTTPListen is the server HTTP listener IP address
	HTTPListener string
	// Value of RecursionAvailable for responses in Mesos domain
	RecurseOn bool
	// Enable serving DSN and HTTP requests
	DNSOn  bool `json:"DnsOn"`
	HTTPOn bool `json:"HttpOn"`
	// Enable replies for external requests
	ExternalOn bool
	// EnforceRFC952 will enforce an older, more strict set of rules for DNS labels
	EnforceRFC952 bool
	// SetTruncateBit when `false` ensures responses never have the Truncate bit set even
	// if they were truncated. When `true` any message that gets truncated will have the
	// Truncate bit set.
	// Compliant clients that receive truncated responses will retry the request
	// over TCP. Some non-compliant clients simply cannot handle truncated responses at all.
	// Disabling this option causes mesos-dns to behave in a non-compliant way. It exists
	// only as a workaround for non-compliant clients and users who cannot tolerate the
	// latency added by the standard TCP fallback.
	SetTruncateBit bool
	// Enumeration enabled via the API enumeration endpoint
	EnumerationOn bool
	// Communicate with Mesos using HTTPS if set to true
	MesosHTTPSOn bool
	// CA certificate to use to verify Mesos Master certificate
	CACertFile string

	// Client certificate to use.
	CertFile string
	// Client certificate key to use.
	KeyFile string

	MesosCredentials basic.Credentials
	// IAM Config File
	IAMConfigFile string

	caPool *x509.CertPool

	cert tls.Certificate

	httpConfigMap httpcli.ConfigMap

	MesosAuthentication httpcli.AuthMechanism
}

// NewConfig return the default config of the resolver
func NewConfig() Config {
	return Config{
		ZkDetectionTimeout: 30,
		RefreshSeconds:     60,
		TTL:                60,
		SRVRecordDefaultWeight: 1,
		Domain:                 "mesos",
		Port:                   53,
		Timeout:                5,
		StateTimeoutSeconds:    300,
		SOARname:               "root.ns1.mesos",
		SOAMname:               "ns1.mesos",
		SOARefresh:             60,
		SOARetry:               600,
		SOAExpire:              86400,
		SOAMinttl:              60,
		ZoneResolvers:          map[string][]string{},
		Resolvers:              []string{"8.8.8.8"},
		Listener:               "0.0.0.0",
		HTTPListener:           "0.0.0.0",
		HTTPPort:               8123,
		DNSOn:                  true,
		HTTPOn:                 true,
		ExternalOn:             true,
		SetTruncateBit:         true,
		RecurseOn:              true,
		IPSources:              []string{"netinfo", "mesos", "host"},
		EnumerationOn:          true,
		MesosAuthentication:    httpcli.AuthNone,
	}
}

// SetConfig instantiates a Config struct read in from config.json
func SetConfig(cjson string) Config {
	c, err := readConfig(cjson)
	if err != nil {
		logging.Error.Fatal(err)
	}
	logging.Verbose.Printf("config loaded from %q", c.File)
	// validate and complete configuration file
	err = validateEnabledServices(c)
	if err != nil {
		logging.Error.Fatalf("service validation failed: %v", err)
	}
	if err = validateMasters(c.Masters); err != nil {
		logging.Error.Fatal(err)
	}

	c.initResolvers()

	if err = validateIPSources(c.IPSources); err != nil {
		logging.Error.Fatalf("IPSources validation failed: %v", err)
	}

	if c.StateTimeoutSeconds <= 0 {
		logging.Error.Fatal("Invalid HTTP Timeout: ", c.StateTimeoutSeconds)
	}

	c.Domain = strings.ToLower(c.Domain)

	err = validateDomainName(c.Domain)
	if err != nil {
		logging.Error.Fatalf("%s is not a valid domain name", c.Domain)
	}

	c.initSOA()
	c.initCertificates()
	c.initMesosAuthentication()
	c.log()

	return *c
}

func (c *Config) initCertificates() {
	if c.CACertFile != "" {
		pool, err := readCACertFile(c.CACertFile)
		if err != nil {
			logging.Error.Fatal(err.Error())
		}
		c.caPool = pool
	}

	if c.CertFile != "" && c.KeyFile == "" {
		logging.Error.Fatalf("Missing private key")
	}

	if c.CertFile == "" && c.KeyFile != "" {
		logging.Error.Fatalf("Missing certificate")
	}

	if c.CertFile != "" && c.KeyFile != "" {
		cert, err := tls.LoadX509KeyPair(c.CertFile, c.KeyFile)
		if err != nil {
			logging.Error.Fatal(err.Error())
		}
		c.cert = cert
	}
}

func (c *Config) initMesosAuthentication() {
	configMapOpts := httpcli.ConfigMapOptions{
		basic.Configuration(c.MesosCredentials),
	}
	if c.IAMConfigFile != "" {
		iamConfig, err := iam.LoadFromFile(c.IAMConfigFile)
		if err != nil {
			logging.Error.Fatal(err.Error())
		}
		configMapOpts = append(configMapOpts, iam.Configuration(iamConfig))
	}

	c.httpConfigMap = configMapOpts.ToConfigMap()
	err := httpcli.Validate(c.MesosAuthentication, c.httpConfigMap)
	if err != nil {
		logging.Error.Fatal(err.Error())
	}
}

func (c *Config) initResolvers() {
	if c.ExternalOn {
		if len(c.Resolvers) == 0 {
			c.Resolvers = GetLocalDNS()
		}
		if err := validateResolvers(c.Resolvers); err != nil {
			logging.Error.Fatal(err)
		}
		if err := validateZoneResolvers(c.ZoneResolvers, c.Domain); err != nil {
			logging.Error.Fatal(err)
		}
	}
}

func (c *Config) initSOA() {
	// SOA record fields
	c.SOARname = strings.TrimRight(strings.Replace(c.SOARname, "@", ".", -1), ".") + "."
	c.SOAMname = strings.TrimRight(c.SOAMname, ".") + "."
	c.SOASerial = uint32(time.Now().Unix())
}

func (c Config) log() {
	// print configuration file
	zoneResolversJSON, err := json.Marshal(c.ZoneResolvers)
	if err != nil {
		zoneResolversJSON = []byte(fmt.Sprintf("error: %v", err))
	}
	logging.Verbose.Println("Mesos-DNS configuration:")
	logging.Verbose.Println("   - Masters: " + strings.Join(c.Masters, ", "))
	logging.Verbose.Println("   - Zookeeper: ", c.Zk)
	logging.Verbose.Println("   - ZookeeperDetectionTimeout: ", c.ZkDetectionTimeout)
	logging.Verbose.Println("   - RefreshSeconds: ", c.RefreshSeconds)
	logging.Verbose.Println("   - Domain: " + c.Domain)
	logging.Verbose.Println("   - Listener: " + c.Listener)
	logging.Verbose.Println("   - HTTPListener: " + c.HTTPListener)
	logging.Verbose.Println("   - Port: ", c.Port)
	logging.Verbose.Println("   - DnsOn: ", c.DNSOn)
	logging.Verbose.Println("   - TTL: ", c.TTL)
	logging.Verbose.Println("   - SRVRecordDefaultWeight: ", c.SRVRecordDefaultWeight)
	logging.Verbose.Println("   - Timeout: ", c.Timeout)
	logging.Verbose.Println("   - StateTimeoutSeconds: ", c.StateTimeoutSeconds)

	logging.Verbose.Println("   - ZoneResolvers: " + string(zoneResolversJSON))
	logging.Verbose.Println("   - Resolvers: " + strings.Join(c.Resolvers, ", "))
	logging.Verbose.Println("   - ExternalOn: ", c.ExternalOn)
	logging.Verbose.Println("   - SOAMname: " + c.SOAMname)
	logging.Verbose.Println("   - SOARname: " + c.SOARname)
	logging.Verbose.Println("   - SOASerial: ", c.SOASerial)
	logging.Verbose.Println("   - SOARefresh: ", c.SOARefresh)
	logging.Verbose.Println("   - SOARetry: ", c.SOARetry)
	logging.Verbose.Println("   - SOAExpire: ", c.SOAExpire)
	logging.Verbose.Println("   - SOAExpire: ", c.SOAMinttl)
	logging.Verbose.Println("   - RecurseOn: ", c.RecurseOn)
	logging.Verbose.Println("   - HttpPort: ", c.HTTPPort)
	logging.Verbose.Println("   - HttpOn: ", c.HTTPOn)
	logging.Verbose.Println("   - ConfigFile: ", c.File)
	logging.Verbose.Println("   - EnforceRFC952: ", c.EnforceRFC952)
	logging.Verbose.Println("   - SetTruncateBit: ", c.SetTruncateBit)
	logging.Verbose.Println("   - IPSources: ", c.IPSources)
	logging.Verbose.Println("   - EnumerationOn", c.EnumerationOn)
	logging.Verbose.Println("   - MesosHTTPSOn", c.MesosHTTPSOn)
	logging.Verbose.Println("   - CACertFile", c.CACertFile)
	logging.Verbose.Println("   - CertFile", c.CertFile)
	logging.Verbose.Println("   - KeyFile", c.KeyFile)
	logging.Verbose.Println("   - MesosAuthentication: ", c.MesosAuthentication)
	switch c.MesosAuthentication {
	case httpcli.AuthBasic:
		logging.Verbose.Println("   - MesosCredentials: ", c.MesosCredentials.Principal+":******")
	case httpcli.AuthIAM:
		logging.Verbose.Println("   - IAMConfigFile", c.IAMConfigFile)
	case httpcli.AuthNone:
		if c.MesosCredentials.Principal != "" {
			logging.Error.Println("Warning! MesosCredentials is configured, but " +
				"MesosAuthentication is set to none. This is probably not intentional")
		}
		if c.IAMConfigFile != "" {
			logging.Error.Println("Warning! IAMConfigFile is set, but " +
				"MesosAuthentication is set to none. This is probably not intentional")
		}
	}
}

func readCACertFile(caCertFile string) (caPool *x509.CertPool, err error) {
	var f *os.File
	if f, err = os.Open(caCertFile); err != nil { // nolint: gosec
		err = fmt.Errorf("CACertFile open failed: %v", err)
		return
	}
	defer errorutil.Ignore(f.Close)

	var b []byte
	if b, err = ioutil.ReadAll(f); err != nil {
		err = fmt.Errorf("CACertFile read failed: %v", err)
		return
	}

	pool := x509.NewCertPool()
	if !pool.AppendCertsFromPEM(b) {
		err = fmt.Errorf("CACertFile parsing failed: %v", err)
	} else {
		caPool = pool
	}
	return
}

func readConfig(file string) (*Config, error) {
	c := NewConfig()

	workingDir := "."
	for _, name := range []string{"HOME", "USERPROFILE"} { // *nix, windows
		if dir := os.Getenv(name); dir != "" {
			workingDir = dir
		}
	}

	var err error
	c.File, err = filepath.Abs(strings.Replace(file, "~/", workingDir+"/", 1))
	if err != nil {
		return nil, fmt.Errorf("cannot find configuration file")
	} else if bs, err := ioutil.ReadFile(c.File); err != nil {
		return nil, fmt.Errorf("missing configuration file: %q", c.File)
	} else if err = json.Unmarshal(bs, &c); err != nil {
		return nil, fmt.Errorf("failed to unmarshal config file %q: %v", c.File, err)
	}

	return &c, nil
}

func unique(ss []string) []string {
	set := make(map[string]struct{}, len(ss))
	out := make([]string, 0, len(ss))
	for _, s := range ss {
		if _, ok := set[s]; !ok {
			set[s] = struct{}{}
			out = append(out, s)
		}
	}
	return out
}

// GetLocalDNS returns the first nameserver in /etc/resolv.conf
// Used for non-Mesos queries.
func GetLocalDNS() []string {
	conf, err := dns.ClientConfigFromFile("/etc/resolv.conf")
	if err != nil {
		logging.Error.Fatalf("%v", err)
	}

	return nonLocalAddies(conf.Servers)
}

// Returns non-local nameserver entries
func nonLocalAddies(cservers []string) []string {
	bad := localAddies()

	good := []string{}

	for i := 0; i < len(cservers); i++ {
		local := false
		for x := 0; x < len(bad); x++ {
			if cservers[i] == bad[x] {
				local = true
			}
		}

		if !local {
			good = append(good, cservers[i])
		}
	}

	return good
}

// Returns an array of local ipv4 addresses
func localAddies() []string {
	addies, err := net.InterfaceAddrs()
	if err != nil {
		logging.Error.Println(err)
	}

	bad := []string{}

	for i := 0; i < len(addies); i++ {
		ip, _, err := net.ParseCIDR(addies[i].String())
		if err != nil {
			logging.Error.Println(err)
		}
		// The servers in cservers above could include
		// ipv6, so we should include local ipv6 for
		// that filtering
		if t4 := ip.To4(); t4 != nil {
			bad = append(bad, t4.String())
		} else if t6 := ip.To16(); t6 != nil {
			bad = append(bad, t6.String())
		}
	}

	return bad
}
